### PR TITLE
Persistent tasks

### PR DESCRIPTION
**Pull request by stgleb**
Add persistence for tasks

### CODE REVIEW

File: pkg/workflows/workflow.go [hunk 1]
```diff
@@ -2,27 +2,93 @@ package workflows
...
+
+type Synchronizer interface {
+	Sync([]byte) error
 }
 
-func New(steps []steps.Step, config steps.Config) *WorkFlow {
-	return &WorkFlow{
-		s:      steps,
-		config: config,
+func New(workflowType string, config steps.Config, syncer Synchronizer) (*WorkFlow, error) {
```
Comment (egopher) [inline pkg/workflows/workflow.go]: I don't like that master/node workflow. A user would be able to create his/her own workflows. Better make a field like name.

Comment (stgleb) [reply pkg/workflows/workflow.go]: @egopher In future we can add something like workflow builder, but now we need a minimum workable set of workflows for cloud providers. For building workflows it is better to have template files for that, but this is stretch goal.

File: pkg/workflows/steps/step.go [hunk 2]
```diff
@@ -2,8 +2,35 @@ package steps
...
 
 type Step interface {
-	Run(ctx context.Context, config Config) error
+	Run(context.Context, io.Writer, Config) error
+	Name() string
+	Description() string
+}
+
+var (
+	stepMap map[string]Step
```
Comment (egopher) [inline pkg/workflows/steps/step.go]: Why do we need that step map?

Comment (stgleb) [reply pkg/workflows/steps/step.go]: @egopher Just a single centralized storage for steps. Each subpackage creates it own step and register itself in that map.

File: pkg/templatemanager/template_manager.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	"strings"
+	"text/template"
+)
+
+var (
+	templateMap map[string]*template.Template
+)
+
+func Init(templateDir string) error {
+	files, err := ioutil.ReadDir(templateDir)
```
Comment (saheienko) [inline pkg/templatemanager/template_manager.go]: It's a bad idea to depend on an external directory. Could we embed these files to the binary?

File: pkg/workflows/steps/certificates/write_certificates.go [hunk 4]
```diff
@@ -7,40 +7,43 @@ import (
...
 
-	"github.com/supergiant/supergiant/pkg/runner"
-	"github.com/supergiant/supergiant/pkg/runner/ssh"
+	tm "github.com/supergiant/supergiant/pkg/templatemanager"
 	"github.com/supergiant/supergiant/pkg/workflows/steps"
 )
 
-type Task struct {
-	runner runner.Runner
+const StepName = "write_certificates"
```
Comment (saheienko) [inline pkg/workflows/steps/certificates/write_certificates.go]: Could we use camelCase names here?

File: pkg/workflows/steps/cni/cni.go [hunk 5]
```diff
@@ -7,40 +7,42 @@ import (
...
 }
 
-func (j *Task) Run(ctx context.Context, config steps.Config) error {
-	err := steps.RunTemplate(ctx, j.script, j.runner, j.output, nil)
+func (j *Step) Run(ctx context.Context, out io.Writer, config steps.Config) error {
+	err := steps.RunTemplate(ctx, j.script, config.Runner, out, nil)
 
 	if err != nil {
-		return errors.Wrap(err, "error running cni template as a command")
+		return errors.Wrap(err, "error running cni templatemanager as a command")
```
Comment (saheienko) [inline pkg/workflows/steps/cni/cni.go]: Is cni a templatemanager?

File: pkg/workflows/steps/digitalocean/digital_ocean.go [hunk 6]
```diff
@@ -12,11 +12,15 @@ import (
 	"github.com/pkg/errors"
 	"golang.org/x/oauth2"
 
+	"io"
```
Comment (saheienko) [inline pkg/workflows/steps/digitalocean/digital_ocean.go]: Please, apply goimports to this file.

File: pkg/workflows/steps/flannel/flannel_test.go [hunk 7]
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/pkg/errors"
 
 	"context"
+
```
Comment (saheienko) [inline pkg/workflows/steps/flannel/flannel_test.go]: Please, apply goimports to this file.

File: pkg/workflows/workflow.go [hunk 8]
```diff
@@ -2,27 +2,196 @@ package workflows
...
+	"github.com/supergiant/supergiant/pkg/workflows/steps/kubeproxy"
+	"github.com/supergiant/supergiant/pkg/workflows/steps/poststart"
+	"github.com/supergiant/supergiant/pkg/workflows/steps/systemd"
+	"github.com/supergiant/supergiant/pkg/workflows/steps/tiller"
 )
 
+type StepStatus struct {
+	Status   steps.Status `json:"status"`
+	StepName string       `json:"step_name"`
+	ErrMsg   string       `json:"error_message"`
```
Comment (saheienko) [inline pkg/workflows/workflow.go]: Please, use a camelCase format for json tags.

File: pkg/workflows/steps/cni/cni.go (hunk 5 above)
Comment (stgleb) [reply pkg/workflows/steps/cni/cni.go]: No, it was a dumb IDE refactoring

File: pkg/templatemanager/template_manager.go (hunk 3 above)
Comment (stgleb) [reply pkg/templatemanager/template_manager.go]: Yeah, but in that case we cannot allow user to provide his own templates. Anyway this sounds mostly like extend. Templates are itself out of scope for this task.

File: pkg/workflows/workflow.go [hunk 9]
```diff
@@ -2,27 +2,196 @@ package workflows
...
+		steps.GetStep(digitalocean.StepName),
+		steps.GetStep(docker.StepName),
+		steps.GetStep(kubelet.StepName),
+		steps.GetStep(kubeletconf.StepName),
+		steps.GetStep(kubeproxy.StepName),
+		steps.GetStep(systemd.StepName),
+		steps.GetStep(flannel.StepName),
+		steps.GetStep(certificates.StepName),
+		steps.GetStep(cni.StepName),
+		steps.GetStep(tiller.StepName),
```
Comment (saheienko) [inline pkg/workflows/workflow.go]: There is no need in tiller on nodes. Move over, this task should be run once.

File: pkg/workflows/workflow.go [hunk 10]
```diff
@@ -2,27 +2,196 @@ package workflows
...
+
+var (
+	digitalOceanMaster = []steps.Step{
+		steps.GetStep(digitalocean.StepName),
+		steps.GetStep(docker.StepName),
+		steps.GetStep(kubelet.StepName),
+		steps.GetStep(kubeletconf.StepName),
+		steps.GetStep(kubeproxy.StepName),
+		steps.GetStep(systemd.StepName),
+		steps.GetStep(flannel.StepName),
```
Comment (saheienko) [inline pkg/workflows/workflow.go]: Flannel should start before docker.

**Comment by saheienko** [comment]
Please, review all the error messages and imports formatting.

File: pkg/templatemanager/template_manager.go (hunk 3 above)
Comment (egopher) [reply pkg/templatemanager/template_manager.go]: @saheienko we will store them in ETCD in the future

Comment (saheienko) [reply pkg/templatemanager/template_manager.go]: I've got it.

File: pkg/workflows/workflow.go (hunk 9 above)
Comment (stgleb) [reply pkg/workflows/workflow.go]: Ok

File: pkg/workflows/steps/step.go (hunk 2 above)
Comment (saheienko) [reply pkg/workflows/steps/step.go]: Please, use mutex for it.

**Pull request merged**